iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -50,179 +71,112 @@ fn do_rsync<'a>(
-            );
-        } else {
-            rsync_list.push(Work {
-                fid,
+                " moved {} to {} {:?}",
                 src_file,
                 dest_file,
-            });
-
-            if rsync_list.len() >= 500 {
-                let mut rsync_res = do_rsync(&rsync_list).await?;
-
-                result.append(&mut rsync_res);
-                rsync_res.clear();
-                rsync_list.clear();
-            }
-        }
-    }
-
-    let mut rsync_res = do_rsync(&rsync_list).await?;
-
-    result.append(&mut rsync_res);
-
-    Ok(result)
-}
-
-async fn restore_fids(
-    llapi: LlapiFid,
-    dest_src: &str,
-    fid_list: Vec<FidItem>,
-) -> Result<Vec<FidError>, ImlAgentError> {
-    let mut result = vec![];
-    let mpi_path = format!("{}/mpirun", env::get_openmpi_path());
-    let mpi_count = env::get_openmpi_count();
-
-    for fid in fid_list {
-        let fid_path = llapi.fid2path(&fid.fid)?;
-        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
-        let dest_file = format!("{}/{}", &dest_src, &fid_path);
-        let mut dest_dir = PathBuf::from(&src_file);
-
-        dest_dir.pop();
-        fs::create_dir_all(&dest_dir).await?;
-
-        let md = fs::metadata(&dest_file).await?;
-        /* invoking mpifileutils is slow, so only do it for directories
-         * and big files, otherwise just use rsync
-         */
-        let output;
-        if md.is_dir() || (md.len() > LARGEFILE) {
-            output = Command::new(&mpi_path)
-                .arg("--allow-run-as-root")
-                .arg("-c")
-                .arg(format!("{}", mpi_count))
-                .arg("--hostfile")
-                .arg("/etc/iml/filesync-hostfile")
-                .arg("dsync")
-                .arg("-S")
-                .arg(dest_file)
-                .arg(src_file)
-                .kill_on_drop(true)
-                .output();
-        } else {
-            output = Command::new("rsync")
-                .arg(dest_file)
-                .arg(src_file)
-                .kill_on_drop(true)
-                .output();
-        }
-
-        let output = output.await?;
+                output.status.code()
+            );
 
-        if !output.status.success() {
-            // TODO: handle termination conditions without defaulting to 0
-            let res = FidError {
-                fid: fid.fid.clone(),
-                data: fid.data.clone(),
+            Ok::<_, ImlAgentError>(FidError {
+                fid: work.fid.fid.clone(),
+                data: work.fid.data.clone(),
                 errno: output.status.code().unwrap_or(1) as i16,
-            };
-            result.push(res);
-        }
-        tracing::debug!(
-            "exited with {} {} {}",
-            output.status,
-            std::str::from_utf8(&output.stdout)?,
-            std::str::from_utf8(&output.stderr)?
-        );
-    }
+            })
+        })
+        .chunks(1)
```

Comment:
`.chunks(1)` is the same as doing things serially

Reply:
Yes.  dsync is the mpifileutils version of rsync.  It will consume *all* the resources across *all* the clients associated with it.  I really, really want this to be serial.

Reply:
Sure. I just mean `.chunks(1)` means the combinator is not necessary. I'll post a version of this fn that is serial.

Reply:
Use this instead:

```rust
fn do_dsync<'a>(
    src_root: &'a PathBuf,
    dest_root: &'a PathBuf,
    work_list: Vec<Work>,
) -> impl Future<Output = Result<Vec<FidError>, ImlAgentError>> + 'a {
    stream::iter(work_list)
        .map(Ok)
        .and_then(
            move |Work {
                      file_path,
                      fid: FidItem { fid, data },
                  }| async move {
                let mpi_path = format!("{}/mpirun", env::get_openmpi_path());
                let mpi_count = env::get_openmpi_count();
                let src_file = src_root.join(&file_path);
                let dest_file = dest_root.join(&file_path);

                let mut dest_dir = PathBuf::from(&src_file);
                dest_dir.pop();
                fs::create_dir_all(&dest_dir).await?;

                let output = Command::new(mpi_path)
                    .arg("--allow-run-as-root")
                    .arg("-c")
                    .arg(format!("{}", mpi_count))
                    .arg("--hostfile")
                    .arg("/etc/iml/filesync-hostfile")
                    .arg("dsync")
                    .arg("-S")
                    .arg(&src_file)
                    .arg(&dest_file)
                    .kill_on_drop(true)
                    .output()
                    .await?;

                tracing::debug!(
                    " moved {} to {} {:?}",
                    src_file.display(),
                    dest_file.display(),
                    output.status.code()
                );

                Ok::<_, ImlAgentError>(FidError {
                    fid,
                    data,
                    errno: output.status.code().unwrap_or(1) as i16,
                })
            },
        )
        .try_filter(|x| future::ready(x.errno != 0))
        .try_collect()
}
```

This has the same semantics but removes the chunking and collection into a vec with one element, which is not necessary.